packages/core/src/render3/after_render_hooks.ts

```diff
@@ -0,0 +1,212 @@
+ *       console.log('content height: ' + this.contentRef.nativeElement.scrollHeight);
+ *     });
+ *   }
+ * }
+ * ```
+ *
+ * @developerPreview
+ */
+export function afterRender(callback: VoidFunction): AfterRenderHookRef {
+  ngDevMode && assertInInjectionContext(afterRender);
+
+  if (!isPlatformBrowser()) {
+    return {destroy() {}};
+  }
+
+  let destroy: VoidFunction|undefined;
+  const unregisterFn = inject(DestroyRef).onDestroy(() => destroy?.());
+  const manager = inject(AfterRenderHooksManager);
+  const hook = new AfterRenderHook(callback);
+
+  destroy = () => {
+    manager.unregister(hook);
+    unregisterFn();
+  };
+  manager.register(hook);
+  return {destroy};
+}
+
+/**
+ * Register a callback to be invoked the next time the application
+ * finishes rendering.
+ *
+ * Note that the callback will run
+ * - in the order it was registered
+ * - on browser platforms only
+ *
+ * @param callback A callback function to register
+ *
+ * @usageNotes
+ *
+ * Use `afterNextRender` to read or write the DOM once,
+ * for example to initialize a non-Angular library.
+ *
+ * ### Example
+ * ```ts
+ * @Component({
+ *   selector: 'my-chart-cmp',
+ *   template: `<div #chart>{{ ... }}</div>`,
+ * })
+ * export class MyChartCmp {
+ *   @ViewChild('chart') chartRef: ElementRef;
+ *   chart: MyChart|null;
+ *
+ *   constructor() {
+ *     afterNextRender(() => {
+ *       this.chart = new MyChart(this.chartRef.nativeElement);
+ *     });
+ *   }
+ * }
+ * ```
+ *
+ * @developerPreview
+ */
+export function afterNextRender(callback: VoidFunction): AfterRenderHookRef {
+  ngDevMode && assertInInjectionContext(afterNextRender);
+
+  if (!isPlatformBrowser()) {
+    return {destroy() {}};
+  }
+
+  let destroy: VoidFunction|undefined;
+  const unregisterFn = inject(DestroyRef).onDestroy(() => destroy?.());
+  const manager = inject(AfterRenderHooksManager);
+  const hookInvokedOnce = new AfterRenderHook(() => {
+    destroy?.();
+    callback();
+  });
+
+  destroy = () => {
+    manager.unregister(hookInvokedOnce);
+    unregisterFn();
+  };
+  manager.register(hookInvokedOnce);
+  return {destroy};
+}
+
+/**
+ * A wrapper around a function to be used as an after render hook.
+ */
+class AfterRenderHook {
+  private callback: VoidFunction;
+
+  constructor(callback: VoidFunction) {
+    this.callback = callback;
+  }
+
+  invoke() {
+    this.callback();
+  }
+}
+
+/**
+ * Implements `afterRender` callback manager logic.
+ */
+export class AfterRenderHooksManager {
+  private hooks: Set<AfterRenderHook> = new Set();
+  private renderDepth: number = 0;
+  private runningCallbacks: boolean = false;
+
+  /**
+   * Mark the beginning of a render operation (i.e. CD cycle).
+   * Throws if called from an `afterRender` callback.
+   */
+  begin() {
+    if (this.runningCallbacks) {
+      throw new RuntimeError(
+          RuntimeErrorCode.RECURSIVE_APPLICATION_RENDER,
+          'A render operation began before the previous operation ended.');
```

Comment:
Let's add a dev mode flag here to tree-shake away the error message:
```suggestion
      throw new RuntimeError(
          RuntimeErrorCode.RECURSIVE_APPLICATION_RENDER,
          ngDevMode && 'A render operation began before the previous operation ended.');
```

Reply:
I think it'd be also great to provide a bit more info here to mention that this change detection was likely triggered from the `afterRender` or `afterNextRender` callbacks (+ any additional info that would help developers to understand where the problem is coming from and how to fix it). Since we'll have the `ngDevMode` flag here, the error message size would not affect the bundle.

Reply:
Ah good catch on `ngDevMode`. I had it there originally, not sure where it went. Thanks :)